### ISSUE TITLE
Bad kerning in `Text()`

### ISSUE THREAD

**Issue by prism0x** [ManimCommunity/manim#2844]
## Description of bug / unexpected behavior

When using `Text()` or `Paragraph()`, the spacing between letters is generally off regardless of the choice of the font. This is probably an issue with Pango.

See the attached images.

## Expected behavior

The spacing between the letters should be more balanced and evenly distributed. 

## How to reproduce the issue

<details><summary>Code for reproducing the problem</summary>

```py
class DemoScene(Scene):
    def construct(self):
        paragraph = Paragraph(
            """
        Lorem ipsum dolor sit amet, consectetur adipiscing elit,
        sed do eiusmod tempor incididunt ut labore et dolore
        magna aliqua. Ut enim ad minim veniam, quis nostrud.""", 
            font="Helvetica", font_size=32, line_spacing = 0.5
        )
        self.add(paragraph)
```

</details>

## Additional media files

<details><summary>Images/GIFs</summary>

![image](https://user-images.githubusercontent.com/93129268/175085934-27bcf9b4-2ea3-4a79-bace-ae1b0798469f.png)
![image](https://user-images.githubusercontent.com/93129268/175088197-f2cfca93-1491-46fd-ba3c-27de0dfc3cc2.png)

</details>

## Logs
<details><summary>Terminal output</summary>

```
PASTE HERE OR PROVIDE LINK TO https://pastebin.com/ OR SIMILAR
```

</details>

## System specifications

<details><summary>System Details</summary>

I rendered it using the Manimator in the Manim Community Server. The version is 0.15.2

- OS (with version, e.g., Windows 10 v2004 or macOS 10.15 (Catalina)):
- RAM:
- Python version (`python/py/python3 --version`):
- Installed modules (provide output from `pip list`):
```
PASTE HERE
```
</details>

<details><summary>LaTeX details</summary>

+ LaTeX distribution (e.g. TeX Live 2020):
+ Installed LaTeX packages:

</details>

<details><summary>FFMPEG</summary>

Output of `ffmpeg -version`:

```
PASTE HERE
```
</details>

**Comment by Fairlight8** [comment]
I've found this same issue. I think it's a problem about pango and/or cairo and the font size. Let me explain.

## Cause of the issue

If you run this code:

    class BadKerning(Scene):
        def construct(self):
            import manimpango
  
              g = VGroup()
              for size in range(0,100,2):
                  g += Text("{}:THIN".format(size), font="Open Sans", font_size=size, disable_ligatures=True)
              self.add(g.arrange(DOWN).scale(0.25))

You will something like this:

![kerning](https://user-images.githubusercontent.com/36537992/209467725-3bb12885-1e23-4cee-8cf3-0afda381ed9a.jpg)

Using large font sizes, the kerning is good. But if you download the image (it's a 5000x5000), and check smaller font sizes, then you can see bad kerning. I've read: https://gitlab.gnome.org/GNOME/pango/-/issues/686. And I think it's an issue about generating text for screens, in pixel size.

## Proposed solution

Using bigger font sizes clearly improves the kerning. But I don't want to have bigger fonts all the time, obviously. That's why I changed the line 77 in text -> text_mobject.py

`TEXT_MOB_SCALE_FACTOR = 0.005 # Or even a smaller value.`

That reduces the scale or the text object. You have to choose bigger font sizes to look similar, but it will look nice even in small texts.

**Comment by shokirovw** [comment]
@Fairlight8 works fine, thank you for your answer

**Comment by Rapsssito** [comment]
Just in case there are people who are not interested in going directly to modify the source code, this can serve as an alternative:
```python
# Fix for bad kerning in the text
TEXT_SCALE_FACTOR = 0.3

class ScaledMarkupText(MarkupText):
    def __init__(self, *args, **kwargs):
        scale_font = False
        # If the font size is lower than 32, scale it up
        if "font_size" in kwargs and kwargs["font_size"] < 32:
            scale_font = True
            kwargs["font_size"] /= TEXT_SCALE_FACTOR
        super().__init__(*args, **kwargs)
        if scale_font:
            self.scale(TEXT_SCALE_FACTOR)
```

It is not perfect, since a small `TEXT_SCALE_FACTOR` may cause the text to line break at random points.

**Comment by Fairlight8** [comment]
Hello @Rapsssito ! Thanks for the addition, I also reached a similar solution, since I had to work with manim in different computers, and having to modify the code in different python environments was a nightmare. Instead of what you propose, I do something more simple: Use a huge font size (for example: original font size * 10) and then scale the mobject itself (.scale(0.1)).

**Comment by blackedout01** [comment]
I don't think this is just an issue with smaller fonts. The following example shows that there is a letter spacing problem, which looks like a float gets transformed to an integer somewhere.
```python
class TextIssue(Scene):
    def construct(self):
        font_size = ValueTracker(56)
        text = always_redraw(lambda: Text(f"Test 123\nNewline", font="Arial", font_size=font_size.get_value(), disable_ligatures=True).move_to(ORIGIN, aligned_edge=UL))
        value = always_redraw(lambda: Text(f"{font_size.get_value():g}").to_corner(UL))
        self.add(text, value)
        self.play(font_size.animate.set_value(98), run_time=10, rate_func=rate_functions.linear)
        self.wait(0.2)
```

https://github.com/user-attachments/assets/b28d42ee-8184-4c5a-b255-aaa9c63ab512

Generated using Manim Community v0.19.0. The font size animates nicely, but the letter spacing jumps in size even in between integer font sizes. The font size can be seen in the top left corner.

**Comment by uwezi** [comment]
Still this remains a problem in the underlying Pango library and nothing which could easily be affected from within Manim.

**Comment by Fairlight8** [comment]
My best solution is still affecting the scale. Huge font size, and then scale down. But two problems still remain. First, using a huge font size typically forces text to break into new lines all the time. And second, as @blackedout01 said, it is still not perfect, thanks for the animation.

Unfortunately, I started using Latex for generating text. Having a miktex installation is a pain in the ass, but it seems more consistent, so...

**Comment by uwezi** [comment]
I don't think it's relevant here, but why do you think that a MikTeX installation on your system is a "pain in the ass"?

**Comment by meawoppl** [comment]
Another data point: this is strongly **`font_size`-dependent**, and there's a simple workaround.

The uneven letter advances get markedly worse at small `font_size` (roughly ≲ 20). Rendering the *same string at the same final on-screen size*, but computed at a large `font_size` and then scaled down, gives clean spacing:

```python
from manim import *

class TextKerningSize(Scene):
    def construct(self):
        s = "Cassini ranging (Earth-Saturn distance)"
        bad  = Text(s, font_size=14)                 # small -> uneven advances
        good = Text(s, font_size=48).scale(14 / 48)  # identical visual size -> even
        VGroup(bad, good).arrange(DOWN, buff=1.5)
        self.add(bad, good)
```

In the render the `font_size=14` line shows visibly irregular gaps (e.g. `Cassin i  r an gin g`, `distan ce)`), while the scaled-down `font_size=48` line is evenly spaced — at the **same** final size, so it's the layout being computed at the small pixel size rather than the glyphs themselves.

Reproduced on `manim 0.20.1`, Python 3.12.3, Linux (Pango backend).

**Workaround:** build small text at a safe size (~≥ 28px) and `.scale()` it down to the target:

```python
def label(text, font_size, **kw):
    internal = max(font_size, 28)
    t = Text(text, font_size=internal, **kw)
    return t.scale(font_size / internal) if internal != font_size else t
```

Rendered comparison (top = `font_size=14`, bottom = `font_size=48` scaled to the same size):

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/29b115a1-05e2-48aa-816d-2709b8e732d6" />

**Comment by uwezi** [comment]
I don't think we need more data points - the bad kerning for small font sizes in `Text()` is well understood and caused by the Pango library.

**Comment by meawoppl** [comment]
I mainly wanted to share the workaround as it helped me, and hopefully others.

Thanks for a great tool btw.

**Comment by lingyunmo** [comment]
I encountered the exact same issue. The smaller the font size, the more weird gaps appear between the letters. I ended up switching to LaTeX for my text generation and everything is fine now. Just wanted to share this in case it helps others who are still struggling with kerning!

**Comment by Err0r-v2** [comment]
I have a fix for this issue, ill propose it on the ManimPango repository

**Comment by Err0r-v2** [comment]
Here it is : https://github.com/ManimCommunity/ManimPango/issues/130

**Comment by Hun-Bot2** [comment]
A concrete root cause and proposed fix for this has been identified in ManimCommunity/ManimPango#130.

I independently validated the proposed fix on macOS with Pango 1.56.4. Disabling Pango's glyph-position rounding fixes the uneven spacing in my tests, and the resulting glyph positions closely match Manim's Typst text output.

Leaving this here to connect the original issue with the ManimPango investigation.
